Doc/c-api/hash.rst

```diff
@@ -41,8 +45,28 @@ See also the :c:member:`PyTypeObject.tp_hash` member.
    .. versionadded:: 3.4
 
 
+Functions
+^^^^^^^^^
+
+.. c:function:: Py_hash_t PyHash_Double(double value, PyObject *obj)
+
+   Hash a C double number.
+
+   If *value* is not-a-number (NaN):
+
+   * If *obj* is not ``NULL``, return the hash of the *obj* pointer.
+   * Otherwise, return :data:`sys.hash_info.nan <sys.hash_info>` (``0``).
+
+   The function cannot fail: it cannot return ``-1``.
```

Comment:
But we *do* want users to check the result, so that the function can start failing in some cases in the future.
```suggestion
   On failure, the function returns ``-1`` and sets an exception. 
   (``-1`` is not a valid hash value; it is only returned on failure.)
```

Reply:
I am strongly for allowing deprecation via runtime warnings, and for keeping new API consistent in that respect.

If the speed is an issue (which I doubt, with branch prediction around), let's solve that in a way that still allows the API to report errors.

Reply:
> I am strongly for allowing deprecation via runtime warnings, and for keeping new API consistent in that respect.

I created https://github.com/capi-workgroup/api-evolution/issues/43 to discuss functions which cannot fail: when the caller is not expected to check for errors.

> If the speed is an issue (which I doubt, with branch prediction around), let's solve that in a way that still allows the API to report errors.

Would you mind to elaborate how you plan to solve this issue?

My concern is more about usability of the API than performance here.

But yeah, performance matters as well. Such function can be used in a hash table (when floats as used as key), and making such function as fast as possible matters.

Reply:
> Would you mind to elaborate how you plan to solve this issue?

It's possible for specific compilers: add a static inline wrapper with `if (result == -1) __builtin_unreachable();` or `__assume(result != -1)`.
That way the compiler can optimize error checking away, until a later Python version decides to allow failures.